crates/ra_hir_def/src/nameres.rs

```diff
@@ -100,19 +102,77 @@ impl std::ops::Index<LocalModuleId> for CrateDefMap {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+pub enum ModuleOrigin {
+    /// It should not be `None` after collecting definitions.
+    Root(Option<FileId>),
```

Comment:
```suggestion
    CrateRoot {
        definition: FileId
    },
```

Reply:
And let's use `FileId(!0)` as a dummy value, instead of an Option?

Reply:
Really? I would use an explicit dummy value (say, `None`) until one day we can safely initialize it. Though if you're really strong at using `!0`, I can do it anyway.

Reply:
I feel strongly about making sure that, after the initialization, the data structure has the "right" shape. Initialization is a less-interesting stage of life for data, so it's ok if weird stuff happens during init, if that allows for a better interface after initialization is concluded

Reply:
Done